include/mango_v3.hpp

```diff
@@ -259,6 +266,264 @@ struct EventQueue {
+  Uninitialized = 0,
+  InnerNode,
+  LeafNode,
+  FreeNode,
+  LastFreeNode
+};
+
+struct AnyNode {
+  NodeType tag;
+  uint8_t padding[BOOK_NODE_SIZE - 4];
+};
+
+struct InnerNode {
+  NodeType tag;
+  uint32_t prefixLen;
+  __uint128_t key;
+  uint32_t children[2];
+  uint8_t padding[BOOK_NODE_SIZE - 32];
+};
+
+struct LeafNode {
+  NodeType tag;
+  uint8_t ownerSlot;
+  uint8_t orderType;
+  uint8_t version;
+  uint8_t timeInForce;
+  __uint128_t key;
+  solana::PublicKey owner;
+  uint64_t quantity;
+  uint64_t clientOrderId;
+  uint64_t bestInitial;
+  uint64_t timestamp;
+};
+
+struct FreeNode {
+  NodeType tag;
+  uint32_t next;
+  uint8_t padding[BOOK_NODE_SIZE - 8];
+};
+
+struct L1Orderbook {
+  uint64_t highestBid = 0;
+  uint64_t highestBidSize = 0;
+  uint64_t lowestAsk = 0;
+  uint64_t lowestAskSize = 0;
+  double midPoint = 0.0;
+  double spreadBps = 0.0;
+
+  bool valid() const {
+    return ((highestBid && lowestAsk) && (lowestAsk > highestBid)) ? true
+                                                                   : false;
+  }
+};
+
+class NativeToUi {
+ public:
+  NativeToUi(int64_t quoteLotSize, int64_t baseLotSize, uint8_t quoteDecimals,
+             uint8_t baseDecimals) {
+    baseLotsToUiConvertor = baseLotSize / std::pow(10, baseDecimals);
+    priceLotsToUiConvertor = std::pow(10, (baseDecimals - quoteDecimals)) *
+                             quoteLotSize / baseLotSize;
+  }
+
+  double getPrice(int64_t price) const {
+    return price * priceLotsToUiConvertor;
+  }
+
+  double getQuantity(int64_t quantity) const {
+    return quantity * baseLotsToUiConvertor;
+  }
+
+ private:
+  double baseLotsToUiConvertor;
+  double priceLotsToUiConvertor;
+};
+
+class BookSide {
+ public:
+  using order_t = struct LeafNode;
+  using orders_t = std::vector<order_t>;
+
+  BookSide(Side side, uint8_t maxBookDelay = 255)
+      : side(side), maxBookDelay(maxBookDelay) {}
+
+  auto getMaxTimestamp() {
+    const auto now = std::chrono::system_clock::now();
+    const auto nowUnix =
+        std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch())
+            .count();
+
+    auto maxTimestamp = nowUnix - maxBookDelay;
+    auto iter = BookSide::BookSideRaw::iterator(side, *raw);
+    while (iter.stack.size() > 0) {
+      if ((*iter).tag == NodeType::LeafNode) {
+        const auto leafNode =
+            reinterpret_cast<const struct LeafNode*>(&(*iter));
+        if (leafNode->timestamp > maxTimestamp) {
+          maxTimestamp = leafNode->timestamp;
+        }
+      }
+      ++iter;
+    }
+    return maxTimestamp;
+  }
+
+  bool update(const std::string& decoded) {
+    if (decoded.size() != sizeof(BookSideRaw)) {
+      throw std::runtime_error("invalid response length " +
+                               std::to_string(decoded.size()) + " expected " +
+                               std::to_string(sizeof(BookSideRaw)));
+    }
+    memcpy(&(*raw), decoded.data(), sizeof(BookSideRaw));
```

Comment:
needs critical section from before memcpy to end of iterator lifetime.
concurrent update calls else could cause memory corruption.
another option would be to not update the same buffer and to swap references to a new buffer, so that iterator can still iterate over the "old buffer" and de-allocate it on iterator destruction.